types/typemap.go

```diff
@@ -0,0 +1,136 @@
+package types
+
+import (
+	"fmt"
+	"reflect"
+	"unsafe"
+)
+
+// Global type register.
+var Types *TypeMap = NewTypeMap()
+
+// SerializerFn is the signature of custom serializer functions. Use the
+// [Serialize] function to drive the [Serializer]. Returning an error results in
+// the program panicking.
+type SerializerFn[T any] func(*Serializer, *T) error
+
+// DeserializerFn is the signature of customer deserializer functions. Use the
+// [Deserialize] function to drive the [Deserializer]. Returning an error
+// results in the program panicking.
+type DeserializerFn[T any] func(*Deserializer, *T) error
+
+// RegisterSerde attaches custom serialization and deserialization functions to
+// type T.
+//
+// Coroutine state is serialized and deserialized when calling [Context.Marshal]
+// and [Context.Unmarshal] respectively.
+//
+// Go basic types, structs, interfaces, slices, arrays, or any combination of
+// them have built-in serialization and deserialization mechanisms. Channels and
+// sync values do not.
+//
+// Custom serializer and deserializer functions can be attached to types using
+// [RegisterSerde] to control how they are serialized, and possibly perform
+// additional initialization on deserialization. Those functions are drivers for
+// [Serializer] and [Deserializer], that need to invoke [Serialize] and
+// [DeserializeTo] in order to actually perform serialization and
+// deserialization operations. Pointers to the same address are detected as such
+// to be reconstructed as pointing to the same value. Slices are serialized by
+// first serializing their backing array, and then length and capacity. As a
+// result, slices sharing the same backing array are deserialized into one array
+// with two shared slices, just like the original state was. Elements between
+// length and capacity are also preserved.
+func RegisterSerde[T any](
+	serializer SerializerFn[T],
+	deserializer DeserializerFn[T]) {
+	registerSerde[T](Types, serializer, deserializer)
+}
+
+func registerSerde[T any](tm *TypeMap,
+	serializer func(*Serializer, *T) error,
+	deserializer func(*Deserializer, *T) error) {
+
+	t := reflect.TypeOf((*T)(nil)).Elem()
+
+	s := func(s *Serializer, p unsafe.Pointer) {
+		if err := serializer(s, (*T)(p)); err != nil {
+			panic(fmt.Errorf("serializing %s: %w", t, err))
+		}
+	}
+
+	d := func(d *Deserializer, p unsafe.Pointer) {
+		if err := deserializer(d, (*T)(p)); err != nil {
+			panic(fmt.Errorf("deserializing %s: %w", t, err))
+		}
+	}
+
+	tm.Attach(t, s, d)
+}
+
+type serializerFn func(*Serializer, unsafe.Pointer)
+type deserializerFn func(d *Deserializer, p unsafe.Pointer)
```

Comment:
Those are internal types but for consistency, how about we name them `serializeFunc` and `deserializeFunc` as well?